# Camera Sensor Noise Analysis

## Background

https://www.iso.org/obp/ui/#iso:std:iso:12232:ed-3:v1:en

https://www.iso.org/obp/ui/#iso:std:iso:12232:ed-3:v1:en

https://en.wikipedia.org/wiki/Exposure_value

In [1]:
%load_ext autoreload

import sensor_camera
%aimport sensor_camera

GPU acceleration not available (CuPy not installed)


In [2]:
import pandas as pd
import plotly.express as px
import numpy as np
import os
from collections import OrderedDict

In [ ]:
# Create Analysis object with base path
analysis = sensor_camera.Analysis(os.path.join(os.getcwd(), 'detector-noise'))

In [ ]:
# Define scan specifications
scan_specs = OrderedDict([
    ('Epson R-D1', {'path': 'Epson-R-D1', 'suffix': 'ERF'}),
    ('Leica Q (24MP)', {'path': 'Q-24MP', 'suffix': 'DNG'}),
    ('Leica M11 (36MP)', {'path': 'M11-36MP', 'suffix': 'DNG'}),
    ('Leica M11 (60MP)', {'path': 'M11-60MP', 'suffix': 'DNG'}),
    ('Leica M11 (18MP)', {'path': 'M11-18MP', 'suffix': 'DNG'}),
    ('Leica M10 (24MP)', {'path': 'M10-24MP', 'suffix': 'DNG'}),
    ('Leica M Monochrome', {'path': 'M-Monochrome', 'suffix': 'DNG'}),
    ('Ricoh GR III', {'path': 'Ricoh-GRIII', 'suffix': 'DNG'}),
    ('iPhone 13 Pro', {'path': 'iPhone-13-Pro', 'suffix': 'DNG'}),
    ('Leica CL (24MP)', {'path': 'CL-24MP', 'suffix': 'DNG'}),
    ('Leica Q3 (36MP)', {'path': 'Q3-36MP', 'suffix': 'DNG'}),
    ('Leica SL2-S (24MP)', {'path': 'SL2-S-24MP', 'suffix': 'DNG'}),
    ('Ricoh GXR', {'path': 'Ricoh-GXR', 'suffix': 'DNG'}),
    ('Leica TL2 (24MP)', {'path': 'TL2-24MP', 'suffix': 'DNG'}),
    ('Pixii (24MP)', {'path': 'Pixii', 'suffix': 'DNG'}),
])

# Execute scans using Analysis class
scan_results = analysis.scan(scan_specs)


Scanning Epson R-D1...
Loading existing results from: /home/steven/git/camera-char/detector-noise/Epson-R-D1/noise_results.csv
Scanning Leica Q (24MP)...
Loading existing results from: /home/steven/git/camera-char/detector-noise/Q-24MP/noise_results.csv
Scanning Leica M11 (36MP)...
Loading existing results from: /home/steven/git/camera-char/detector-noise/M11-36MP/noise_results.csv
Scanning Leica M11 (60MP)...
Loading existing results from: /home/steven/git/camera-char/detector-noise/M11-60MP/noise_results.csv
Scanning Leica M11 (18MP)...
Loading existing results from: /home/steven/git/camera-char/detector-noise/M11-18MP/noise_results.csv
Scanning Leica M10 (24MP)...
Loading existing results from: /home/steven/git/camera-char/detector-noise/M10-24MP/noise_results.csv
Scanning Leica M Monochrome...
Loading existing results from: /home/steven/git/camera-char/detector-noise/M-Monochrome/noise_results.csv
Scanning Ricoh GR III...
Loading existing results from: /home/steven/git/camera-char/

In [5]:
# Create convenient aliases for backward compatibility
r_d1_6 = scan_results['Epson R-D1']
q_24 = scan_results['Leica Q (24MP)']
m11_36 = scan_results['Leica M11 (36MP)']
m11_60 = scan_results['Leica M11 (60MP)']
m11_18 = scan_results['Leica M11 (18MP)']
m10_24 = scan_results['Leica M10 (24MP)']
m_mono = scan_results['Leica M Monochrome']
gr_iii = scan_results['Ricoh GR III']
iphone_13_pro = scan_results['iPhone 13 Pro']
cl_24 = scan_results['Leica CL (24MP)']
q3_36 = scan_results['Leica Q3 (36MP)']
sl2_s_24 = scan_results['Leica SL2-S (24MP)']
gxr_10 = scan_results['Ricoh GXR']
tl2_24 = scan_results['Leica TL2 (24MP)']
pixii_24 = scan_results['Pixii (24MP)']


## Analyze noise data from cameras

In [ ]:
# Create aggregate data from selected cameras
data = analysis.create_aggregate([
    'Leica M11 (36MP)', 'Leica M10 (24MP)', 'Leica Q (24MP)', 
    'Leica M Monochrome', 'Ricoh GR III', 'Leica CL (24MP)', 
    'Leica Q3 (36MP)', 'Leica SL2-S (24MP)', 'Ricoh GXR', 'Epson R-D1'
])

# Save aggregate analysis to CSV
analysis.save_aggregate('aggregate_analysis.csv')

In [ ]:
# Specify shutter speed (exposure time in seconds)
# Common values: 0.004 (1/250s), 0.001 (1/1000s), 0.008 (1/125s), 0.033 (1/30s)
shutter_speed = 0.001  

fig = analysis.plot_ev_vs_iso(exposure_time=shutter_speed)
fig.show()

## Noise reports for each camera

In [88]:
m11_36[m11_36['time']==0.004]

camera                                             source  \
8    LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
21   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
34   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
47   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
60   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
73   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
86   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
99   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
112  LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
125  LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   

     black_level  white_level  width  height    iso   time         std  \
8           1023        16383   7424    4936     64  0.004    0.956960   
21          1023        16383   7424    4936    200  0.004    1.195606   
34          1023        16383   7424    4936    400  0.004    2.148773   
47          1023        16383   7424    4936    800  0.004    3.948657   
60          1023        16383   7424    4936   1600  0.004    6.897731   
73          1023        16383   7424    4936   3200  0.004   12.715045   
86          1023        16383   7424    4936   6400  0.004   23.514644   
99          1023        16383   7424    4936  12500  0.004   45.202484   
112         1023        16383   7424    4936  25000  0.004   89.263262   
125         2047        16383   7424    4936  50000  0.004  174.226691   

            mean   min   max         EV  
8    1022.929697  1001  1046  13.970361  
21   1023.071412   990  1057  13.649148  
34   1023.099960   975  1072  12.803378  
47   1023.203138   942  1122  11.925528  
60   1023.042264   873  1191  11.120769  
73   1023.329588   760  1329  10.238426  
86   1023.101178   515  1525   9.351403  
99   1023.276510     1  2269   8.408560  
112  1024.055478     0  3178   7.426896  
125  2049.735271     0  6786   6.362533

In [89]:
m11_60[m11_60['time']==0.004]

camera                                             source  \
8    LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
21   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
34   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
48   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
61   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
75   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
88   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
101  LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
114  LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
128  LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   

     black_level  white_level  width  height    iso   time         std  \
8           1023        16383   9536    6336     64  0.004    1.334593   
21          1023        16383   9536    6336    200  0.004    1.706675   
34          1023        16383   9536    6336    400  0.004    3.138273   
48          1023        16383   9536    6336    800  0.004    5.805118   
61          1023        16383   9536    6336   1600  0.004   10.200857   
75          1023        16383   9536    6336   3200  0.004   18.823708   
88          1023        16383   9536    6336   6400  0.004   34.798769   
101         1023        16383   9536    6336  12500  0.004   66.761364   
114         1023        16383   9536    6336  25000  0.004  131.732962   
128         2047        16383   9536    6336  50000  0.004  256.508930   

            mean  min   max         EV         std_2  
8    1023.242184  996  1050  13.490490      1.781140  
21   1023.042660  982  1064  13.135702      2.912740  
34   1023.070871  956  1084  12.256920      9.848760  
48   1023.160621  917  1143  11.369565     33.699398  
61   1023.008775  796  1246  10.556272    104.057476  
75   1023.264036  678  1381   9.672412    354.331996  
88   1023.018773  391  1863   8.785926   1210.954351  
101  1023.151221    0  2425   7.845949   4457.079752  
114  1023.679221    0  3752   6.865418  17353.573367  
128  2049.012691    0  7921   5.804490  65796.831320

In [90]:
m11_18[m11_18['time']==0.004]

camera                                             source  \
8    LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
21   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
34   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
47   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
60   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
73   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
86   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
99   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
112  LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   
125  LEICA M11  /home/steven/git/camera-char/detector-noise/M1...   

     black_level  white_level  width  height    iso   time         std  \
8           1023        16383   5280    3506     64  0.004    0.963765   
21          1023        16383   5280    3506    200  0.004    1.205505   
34          1023        16383   5280    3506    400  0.004    2.174287   
47          1023        16383   5280    3506    800  0.004    3.992646   
60          1023        16383   5280    3506   1600  0.004    6.974517   
73          1023        16383   5280    3506   3200  0.004   12.879853   
86          1023        16383   5280    3506   6400  0.004   23.802647   
99          1023        16383   5280    3506  12500  0.004   45.746737   
112         1023        16383   5280    3506  25000  0.004   90.328608   
125         2047        16383   5280    3506  50000  0.004  176.232835   

            mean   min   max         EV  
8    1023.281389  1003  1046  13.960137  
21   1023.084533   988  1058  13.637253  
34   1023.107873   966  1070  12.786348  
47   1023.200719   939  1116  11.909545  
60   1023.050038   866  1156  11.104797  
73   1023.334011   746  1316  10.219846  
86   1023.092075   476  1602   9.333840  
99   1023.304053     0  2539   8.391294  
112  1024.043486     0  3283   7.409780  
125  2049.782265     0  6043   6.346016

In [91]:
m10_24[m10_24['time']==0.004]

camera                                             source  \
8    LEICA M10  /home/steven/git/camera-char/detector-noise/M1...   
21   LEICA M10  /home/steven/git/camera-char/detector-noise/M1...   
34   LEICA M10  /home/steven/git/camera-char/detector-noise/M1...   
47   LEICA M10  /home/steven/git/camera-char/detector-noise/M1...   
60   LEICA M10  /home/steven/git/camera-char/detector-noise/M1...   
73   LEICA M10  /home/steven/git/camera-char/detector-noise/M1...   
86   LEICA M10  /home/steven/git/camera-char/detector-noise/M1...   
99   LEICA M10  /home/steven/git/camera-char/detector-noise/M1...   
112  LEICA M10  /home/steven/git/camera-char/detector-noise/M1...   
125  LEICA M10  /home/steven/git/camera-char/detector-noise/M1...   

     black_level  white_level  width  height    iso   time         std  \
8              0        15000   5984    3992    100  0.004    1.528083   
21             0        15000   5984    3992    200  0.004    1.610966   
34             0        15000   5984    3992    400  0.004    2.127599   
47             0        15000   5984    3992    800  0.004    3.822102   
60             0        15000   5984    3992   1600  0.004    6.860658   
73             0        15000   5984    3992   3200  0.004   13.395494   
86             0        15000   5984    3992   6400  0.004   26.904964   
99             0        15000   5984    3992  12500  0.004   53.688481   
112            0        14000   5984    3992  25000  0.004  140.644126   
125            0        14000   5984    3992  50000  0.004  276.584708   

           mean  min   max         EV  
8      0.915162    0    86  13.260952  
21     0.991002    0    46  13.184749  
34     1.249475    0    87  12.783448  
47     2.257264    0   156  11.938309  
60     4.093311    0   277  11.094328  
73     7.979042    0   784  10.128999  
86    15.936082    0   467   9.122874  
99    31.615088    0  1046   8.126134  
112   89.519195    0  2791   6.637234  
125  177.941345    0  3960   5.661562

In [92]:
q_24[q_24['time']==0.004]

camera                                             source  \
8    LEICA Q (Typ 116)  /home/steven/git/camera-char/detector-noise/Q-...   
20   LEICA Q (Typ 116)  /home/steven/git/camera-char/detector-noise/Q-...   
32   LEICA Q (Typ 116)  /home/steven/git/camera-char/detector-noise/Q-...   
44   LEICA Q (Typ 116)  /home/steven/git/camera-char/detector-noise/Q-...   
57   LEICA Q (Typ 116)  /home/steven/git/camera-char/detector-noise/Q-...   
69   LEICA Q (Typ 116)  /home/steven/git/camera-char/detector-noise/Q-...   
81   LEICA Q (Typ 116)  /home/steven/git/camera-char/detector-noise/Q-...   
93   LEICA Q (Typ 116)  /home/steven/git/camera-char/detector-noise/Q-...   
105  LEICA Q (Typ 116)  /home/steven/git/camera-char/detector-noise/Q-...   
117  LEICA Q (Typ 116)  /home/steven/git/camera-char/detector-noise/Q-...   

     black_level  white_level  width  height    iso   time         std  \
8            512        16383   6120    4016    100  0.004    3.678146   
20           512        16383   6120    4016    200  0.004    2.188392   
32           512        16383   6120    4016    400  0.004    3.574374   
44           512        16383   6120    4016    800  0.004    6.697252   
57           512        16383   6120    4016   1600  0.004   13.060736   
69           512        16383   6120    4016   3200  0.004   26.370818   
81           512        16383   6120    4016   6400  0.004   52.984696   
93           512        16383   6120    4016  12500  0.004   98.368767   
105          512        16383   6120    4016  25000  0.004  194.501976   
117          512        16383   6120    4016  50000  0.004  379.146844   

           mean  min    max         EV  
8    511.310030    2    718  12.075127  
20   511.070735    1    911  12.824234  
32   511.045290    1   1327  12.116415  
44   511.423587   53   2306  11.210536  
57   511.627440    0   1656  10.246941  
69   513.771085    4   2530   9.233235  
81   517.682071    1   5292   8.226602  
93   533.449369    0   9202   7.333977  
105  542.401243    1  16383   6.350464  
117  584.785371    1  16383   5.387493

In [93]:
m_mono[m_mono['time']==0.004]

camera  \
8   LEICA M MONOCHROM (Typ 246)   
21  LEICA M MONOCHROM (Typ 246)   
34  LEICA M MONOCHROM (Typ 246)   
47  LEICA M MONOCHROM (Typ 246)   
60  LEICA M MONOCHROM (Typ 246)   
73  LEICA M MONOCHROM (Typ 246)   
86  LEICA M MONOCHROM (Typ 246)   
99  LEICA M MONOCHROM (Typ 246)   

                                               source  black_level  \
8   /home/steven/git/camera-char/detector-noise/M-...            0   
21  /home/steven/git/camera-char/detector-noise/M-...            0   
34  /home/steven/git/camera-char/detector-noise/M-...            0   
47  /home/steven/git/camera-char/detector-noise/M-...            0   
60  /home/steven/git/camera-char/detector-noise/M-...            0   
73  /home/steven/git/camera-char/detector-noise/M-...            0   
86  /home/steven/git/camera-char/detector-noise/M-...            0   
99  /home/steven/git/camera-char/detector-noise/M-...            0   

    white_level  width  height    iso   time        std       mean  min  max  \
8          3750   5984    4000    320  0.004   0.224151   0.021169    0  217   
21         3750   5984    4000    400  0.004   0.271029   0.020925    0  305   
34         3750   5984    4000    800  0.004   0.563616   0.046060    0  699   
47         3750   5984    4000   1600  0.004   1.723564   1.125489    0  182   
60         3750   5984    4000   3200  0.004   3.527236   2.355802    0  255   
73         3750   5984    4000   6400  0.004   6.947738   4.487237    0  390   
86         3750   5984    4000  12500  0.004  14.403206   9.608023    0  524   
99         3750   5984    4000  25000  0.004  26.450558  16.854272    0  555   

           EV  
8   14.030132  
21  13.756158  
34  12.699890  
47  11.087280  
60  10.054137  
73   9.076131  
86   8.024357  
99   7.147449

In [94]:
gr_iii[gr_iii['time']==0.004]

camera                                             source  \
9    RICOH GR III  /home/steven/git/camera-char/detector-noise/Ri...   
22   RICOH GR III  /home/steven/git/camera-char/detector-noise/Ri...   
35   RICOH GR III  /home/steven/git/camera-char/detector-noise/Ri...   
48   RICOH GR III  /home/steven/git/camera-char/detector-noise/Ri...   
61   RICOH GR III  /home/steven/git/camera-char/detector-noise/Ri...   
74   RICOH GR III  /home/steven/git/camera-char/detector-noise/Ri...   
87   RICOH GR III  /home/steven/git/camera-char/detector-noise/Ri...   
100  RICOH GR III  /home/steven/git/camera-char/detector-noise/Ri...   
114  RICOH GR III  /home/steven/git/camera-char/detector-noise/Ri...   
127  RICOH GR III  /home/steven/git/camera-char/detector-noise/Ri...   
140  RICOH GR III  /home/steven/git/camera-char/detector-noise/Ri...   

     black_level  white_level  width  height    iso   time         std  \
9             64        16315   6112    4064    100  0.004    1.658870   
22           128        16253   6112    4064    200  0.004    1.544673   
35           256        15870   6112    4064    400  0.004    2.601941   
48           512        15352   6112    4064    800  0.004    3.520311   
61          1024        14323   6112    4064   1600  0.004    6.433804   
74          1024        15091   6112    4064   3200  0.004   12.352671   
87          2048        14306   6112    4064   6400  0.004   24.707159   
100         2048        14269   6112    4064  12800  0.004   49.500448   
114         4032        12215   6112    4064  25600  0.004  101.487408   
127         4032        12082   6112    4064  51200  0.004  212.426192   
140         4032        11796   6112    4064  65535  0.004  428.824388   

            mean   min    max         EV  
9      63.673830    36    412  13.258040  
22    128.589339    77    885  13.349710  
35    256.002652   161   1764  12.550964  
48    511.403617   403    754  12.041501  
61   1023.953199   836   1470  11.013358  
74   1022.470103   626   1914  10.153276  
87   2046.629512  1242   4322   8.954579  
100  2044.358853   308  11363   7.947705  
114  4031.260829   299  16383   6.333257  
127  4042.753656     0  16383   5.243955  
140  4059.739067     0  16383   4.178341

In [95]:
iphone_13_pro

camera                                             source  \
0  iPhone14,2 back camera  /home/steven/git/camera-char/detector-noise/iP...   
1  iPhone14,2 back camera  /home/steven/git/camera-char/detector-noise/iP...   
2  iPhone14,2 back camera  /home/steven/git/camera-char/detector-noise/iP...   
3  iPhone14,2 back camera  /home/steven/git/camera-char/detector-noise/iP...   
4  iPhone14,2 back camera  /home/steven/git/camera-char/detector-noise/iP...   
5  iPhone14,2 back camera  /home/steven/git/camera-char/detector-noise/iP...   
6  iPhone14,2 back camera  /home/steven/git/camera-char/detector-noise/iP...   
7  iPhone14,2 back camera  /home/steven/git/camera-char/detector-noise/iP...   

   black_level  white_level  width  height   iso      time           std  \
0            0        65535   4032    3024    64  0.016667  17679.448169   
1            0        65535   4032    3024   100  0.016667  15488.245049   
2            0        65535   4032    3024   200  0.008000   9938.156688   
3            0        65535   4032    3024   400  0.008000   9167.233663   
4            0        65535   4032    3024   800  0.008000   9419.149043   
5            0        65535   4032    3024  1600  0.008000   8776.751815   
6            0        65535   4032    3024  3200  0.001984   7449.201007   
7            0        65535   4032    3024  6400  0.001976   7402.108694   

           mean  min    max        EV  
0  10372.323346    0  65535  1.890192  
1  10562.273059    0  65535  2.081092  
2   6670.797098    0  65535  2.721215  
3   5687.718531    0  65535  2.837707  
4   6114.673293    0  65535  2.798597  
5   5313.392968    0  65535  2.900507  
6   4176.450377    0  65535  3.137108  
7   4174.681609    0  65535  3.146257

In [96]:
iphone_13_pro[iphone_13_pro['time']==0.004]

Empty DataFrame
Columns: [camera, source, black_level, white_level, width, height, iso, time, std, mean, min, max, EV]
Index: []

Epson R-D1, 6 MP Camera

In [97]:
r_d1_6[r_d1_6['time']==0.004]

camera                                             source  black_level  \
9    R-D1  /home/steven/git/camera-char/detector-noise/Ep...           63   
21   R-D1  /home/steven/git/camera-char/detector-noise/Ep...           63   
33   R-D1  /home/steven/git/camera-char/detector-noise/Ep...           63   
45   R-D1  /home/steven/git/camera-char/detector-noise/Ep...           63   

    white_level  width  height   iso   time        std       mean  min   max  \
9          4096   3040    2024   200  0.004   2.649669  63.322883   52   481   
21         4096   3040    2024   400  0.004   3.919151  63.412075   43   974   
33         4096   3040    2024   800  0.004   7.048594  63.479990   29  2105   
45         4096   3040    2024  1600  0.004  13.212764  63.827821    0  4095   

           EV  
9   10.571825  
21  10.007096  
33   9.160302  
45   8.253777

In [98]:
cl_24[cl_24['time']==0.004]

camera                                             source  black_level  \
11   LEICA CL  /home/steven/git/camera-char/detector-noise/CL...          511   
24   LEICA CL  /home/steven/git/camera-char/detector-noise/CL...          511   
38   LEICA CL  /home/steven/git/camera-char/detector-noise/CL...          511   
51   LEICA CL  /home/steven/git/camera-char/detector-noise/CL...          511   
64   LEICA CL  /home/steven/git/camera-char/detector-noise/CL...          511   
77   LEICA CL  /home/steven/git/camera-char/detector-noise/CL...          507   
90   LEICA CL  /home/steven/git/camera-char/detector-noise/CL...          508   
103  LEICA CL  /home/steven/git/camera-char/detector-noise/CL...          506   
116  LEICA CL  /home/steven/git/camera-char/detector-noise/CL...          475   
129  LEICA CL  /home/steven/git/camera-char/detector-noise/CL...          500   

     white_level  width  height    iso   time         std        mean  min  \
11         16383   6120    4016    100  0.004    1.521095  511.548608   61   
24         16383   6120    4016    200  0.004    2.683932  510.753481    1   
38         16383   6120    4016    400  0.004    2.245829  512.100606    0   
51         16383   6120    4016    800  0.004    4.147295  512.090813    4   
64         16383   6120    4016   1600  0.004    7.852681  510.889616    4   
77         16383   6120    4016   3200  0.004   14.684134  508.058249    6   
90         16383   6120    4016   6400  0.004   27.945610  506.458289    1   
103        16383   6120    4016  12500  0.004   53.622947  498.916057    1   
116        16383   6120    4016  25000  0.004  103.502537  475.063261    1   
129        16383   6120    4016  50000  0.004  200.295423  501.529159    1   

       max         EV  
11     745  13.349086  
24     967  12.529848  
38    1670  12.786948  
51    2549  11.902026  
64    5338  10.981011  
77    6119  10.078374  
90   16382   9.149919  
103  16383   8.209872  
116  16383   7.263943  
129  16383   6.309210

In [99]:
q3_36[q3_36['time']==0.004]

camera                                             source  black_level  \
8    LEICA Q3  /home/steven/git/camera-char/detector-noise/Q3...          512   
20   LEICA Q3  /home/steven/git/camera-char/detector-noise/Q3...          512   
32   LEICA Q3  /home/steven/git/camera-char/detector-noise/Q3...          512   
44   LEICA Q3  /home/steven/git/camera-char/detector-noise/Q3...          512   
56   LEICA Q3  /home/steven/git/camera-char/detector-noise/Q3...          512   
68   LEICA Q3  /home/steven/git/camera-char/detector-noise/Q3...          512   
80   LEICA Q3  /home/steven/git/camera-char/detector-noise/Q3...          512   
92   LEICA Q3  /home/steven/git/camera-char/detector-noise/Q3...          512   
104  LEICA Q3  /home/steven/git/camera-char/detector-noise/Q3...          512   
116  LEICA Q3  /home/steven/git/camera-char/detector-noise/Q3...          512   
128  LEICA Q3  /home/steven/git/camera-char/detector-noise/Q3...          512   

     white_level  width  height    iso   time         std        mean  min  \
8          16383   7424    4936  50000  0.004  155.754517  509.993272    0   
20         16383   7424    4936  25000  0.004   82.989504  509.563822    0   
32         16383   7424    4936  12500  0.004   50.697111  512.091618    0   
44         16383   7424    4936   6400  0.004   28.945286  512.583887    0   
56         16383   7424    4936   3200  0.004   14.962634  512.653604    0   
68         16383   7424    4936   1600  0.004    7.772959  512.433278  233   
80         16383   7424    4936    800  0.004    4.178515  512.349566  377   
92         16383   7424    4936    400  0.004    2.434730  512.615436  452   
104        16383   7424    4936    200  0.004    2.670535  512.702710  450   
116        16383   7424    4936    100  0.004    1.797926  512.318202  477   
128         8191   7424    4936     50  0.004    1.444199  512.165523  493   

      max         EV  
8    3776   6.670975  
20   2063   7.579248  
32   1707   8.290274  
44   1455   9.098849  
56   1124  10.050813  
68    852  10.995641  
80    663  11.891115  
92    594  12.670343  
104   587  12.536977  
116   550  13.107772  
128   529  12.376433

In [100]:
sl2_s_24[sl2_s_24['time']==0.004]

camera                                             source  \
8    LEICA SL2-S  /home/steven/git/camera-char/detector-noise/SL...   
21   LEICA SL2-S  /home/steven/git/camera-char/detector-noise/SL...   
34   LEICA SL2-S  /home/steven/git/camera-char/detector-noise/SL...   
47   LEICA SL2-S  /home/steven/git/camera-char/detector-noise/SL...   
60   LEICA SL2-S  /home/steven/git/camera-char/detector-noise/SL...   
73   LEICA SL2-S  /home/steven/git/camera-char/detector-noise/SL...   
86   LEICA SL2-S  /home/steven/git/camera-char/detector-noise/SL...   
99   LEICA SL2-S  /home/steven/git/camera-char/detector-noise/SL...   
112  LEICA SL2-S  /home/steven/git/camera-char/detector-noise/SL...   
125  LEICA SL2-S  /home/steven/git/camera-char/detector-noise/SL...   
138  LEICA SL2-S  /home/steven/git/camera-char/detector-noise/SL...   
151  LEICA SL2-S  /home/steven/git/camera-char/detector-noise/SL...   

     black_level  white_level  width  height    iso   time         std  \
8            512         8191   6048    4048     50  0.004    1.505730   
21           511        16383   6048    4048    100  0.004    1.849275   
34           512        16383   6048    4048    200  0.004    2.373474   
47           512        16383   6048    4048    400  0.004    4.052535   
60           511        16383   6048    4048    800  0.004    2.480478   
73           512        16383   6048    4048   1600  0.004    4.199798   
86           515        16383   6048    4048   3200  0.004    7.998514   
99           515        16383   6048    4048   6400  0.004   15.409505   
112          529        16383   6048    4048  12500  0.004   28.669308   
125          548        16383   6048    4048  25000  0.004   56.221427   
138          577        16383   6048    4048  50000  0.004  108.081454   
151          643        16383   6048    4048  65535  0.004  210.235621   

           mean  min    max         EV  
8    510.819759  500    522  12.316240  
21   510.258311  484    535  13.067236  
34   511.269698  459    555  12.707105  
47   510.601590  405    599  11.935281  
60   510.779836  445    661  12.643578  
73   511.189638  368    802  11.883786  
86   513.970802  257   1123  10.954101  
99   514.648340   19   1867  10.008084  
112  530.041126    1   2958   9.111124  
125  547.780757    1   5658   8.137781  
138  585.337123    1   9539   7.192210  
151  660.334851    1  16383   6.226285

Something is wrong with the SL2-S noise calculations - the EV seems to bottom out at 9.9, which is odd. Maybe take a smaller area? 

Doing a crop to 1000:1500, 1000:1500 gave more reasonable results - need to look back at the supposedly correctly cropped data.

In [101]:
gxr_10[gxr_10['time']==0.004]

camera                                             source  \
8   RICOH GXR MOUNT A12  /home/steven/git/camera-char/detector-noise/Ri...   
21  RICOH GXR MOUNT A12  /home/steven/git/camera-char/detector-noise/Ri...   
34  RICOH GXR MOUNT A12  /home/steven/git/camera-char/detector-noise/Ri...   
49  RICOH GXR MOUNT A12  /home/steven/git/camera-char/detector-noise/Ri...   
62  RICOH GXR MOUNT A12  /home/steven/git/camera-char/detector-noise/Ri...   
77  RICOH GXR MOUNT A12  /home/steven/git/camera-char/detector-noise/Ri...   

    black_level  white_level  width  height   iso   time        std      mean  \
8             0         4095   3792    2846   100  0.004   0.896950  0.398122   
21            0         4095   3792    2846   200  0.004   0.905981  0.406497   
34            0         4095   3792    2846   400  0.004   1.817798  0.834600   
49            0         4095   3792    2846   800  0.004   3.910494  1.705950   
62            0         4095   3792    2846  1600  0.004   7.088816  3.364558   
77            0         4095   3792    2846  3200  0.004  13.169909  7.041692   

    min   max         EV  
8     0   798  12.156548  
21    0   549  12.142095  
34    0  1249  11.137456  
49    0  4095  10.032297  
62    0  4095   9.174103  
77    0  4095   8.280474

In [102]:
tl2_24[tl2_24['time']==0.004]

camera                                             source  \
8    LEICA TL2  /home/steven/git/camera-char/detector-noise/TL...   
21   LEICA TL2  /home/steven/git/camera-char/detector-noise/TL...   
34   LEICA TL2  /home/steven/git/camera-char/detector-noise/TL...   
47   LEICA TL2  /home/steven/git/camera-char/detector-noise/TL...   
60   LEICA TL2  /home/steven/git/camera-char/detector-noise/TL...   
73   LEICA TL2  /home/steven/git/camera-char/detector-noise/TL...   
86   LEICA TL2  /home/steven/git/camera-char/detector-noise/TL...   
99   LEICA TL2  /home/steven/git/camera-char/detector-noise/TL...   
112  LEICA TL2  /home/steven/git/camera-char/detector-noise/TL...   
126  LEICA TL2  /home/steven/git/camera-char/detector-noise/TL...   

     black_level  white_level  width  height    iso   time        std  \
8              0        16383   6000    4000    100  0.004   0.043376   
21             0        16383   6000    4000    200  0.004   0.157860   
34             0        16383   6000    4000    400  0.004   0.352779   
47             0        16383   6000    4000    800  0.004   0.898489   
60             0        16383   6000    4000   1600  0.004   1.952132   
73             0        16383   6000    4000   3200  0.004   4.240475   
86             0        16383   6000    4000   6400  0.004  11.956876   
99             0        16383   6000    4000  12500  0.004  24.895169   
112            0        16383   6000    4000  25000  0.004  46.013025   
126            0        16383   6000    4000  50000  0.004  79.582535   

          mean  min    max         EV  
8     0.000460    0     28  18.526862  
21    0.004987    0     64  16.663196  
34    0.022721    0    116  15.503074  
47    0.135710    0    244  14.154339  
60    0.397780    0    516  13.034861  
73    1.027268    0   1204  11.915686  
86    7.829996    0   2432  10.420143  
99   17.316480    0   6216   9.362118  
112  35.174396    0   1288   8.475942  
126  56.012048    0  15248   7.685532

The noise in the detector is way too low. The mean value is also way too low. Is the TL2 scaling the data to bias it around 0?

In [103]:
pixii_24[pixii_24['time']==0.004]

camera                                             source  \
8    Pixii Camera (A2572)  /home/steven/git/camera-char/detector-noise/Pi...   
21   Pixii Camera (A2572)  /home/steven/git/camera-char/detector-noise/Pi...   
37   Pixii Camera (A2572)  /home/steven/git/camera-char/detector-noise/Pi...   
52   Pixii Camera (A2572)  /home/steven/git/camera-char/detector-noise/Pi...   
65   Pixii Camera (A2572)  /home/steven/git/camera-char/detector-noise/Pi...   
78   Pixii Camera (A2572)  /home/steven/git/camera-char/detector-noise/Pi...   
91   Pixii Camera (A2572)  /home/steven/git/camera-char/detector-noise/Pi...   
105  Pixii Camera (A2572)  /home/steven/git/camera-char/detector-noise/Pi...   
118  Pixii Camera (A2572)  /home/steven/git/camera-char/detector-noise/Pi...   
129  Pixii Camera (A2572)  /home/steven/git/camera-char/detector-noise/Pi...   

     black_level  white_level  width  height    iso   time        std  \
8            126         4095   6288    4200    160  0.004  28.640959   
21           124         4095   6288    4200    320  0.004  28.651491   
37           125         4095   6288    4200    640  0.004  28.671363   
52           122         4095   6288    4200   1280  0.004  28.752533   
65           125         4095   6288    4200   2560  0.004  28.704880   
78           122         4095   6288    4200   5120  0.004  28.873546   
91           122         4095   6288    4200   5120  0.004  28.871677   
105          124         4095   6288    4200   3200  0.004  28.740011   
118          121         4095   6288    4200   6400  0.004  28.993666   
129          114         4095   6288    4200  12800  0.004  29.857583   

           mean  min   max        EV  
8    126.102465    0  4095  7.114552  
21   124.248696    0  4095  7.114748  
37   125.679027    0  4095  7.113385  
52   122.649215    0  4095  7.110396  
65   125.280623    0  4095  7.111699  
78   122.281378    0  4095  7.104337  
91   122.272597    0  4095  7.104430  
105  124.390226    0  4095  7.110298  
118  120.901248    0  4095  7.098710  
129  114.580852    0  4095  7.058890

Something's odd with the Pixii data. Is he taking a diff with the background?

## More Analysis

In [24]:
m11_36[m11_36['iso'] == 64]

camera                                             source  black_level  \
0   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...         1023   
1   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...         1023   
2   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...         1023   
3   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...         1023   
4   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...         1023   
5   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...         1023   
6   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...         1023   
7   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...         1023   
8   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...         1023   
9   LEICA M11  /home/steven/git/camera-char/detector-noise/M1...         1023   
10  LEICA M11  /home/steven/git/camera-char/detector-noise/M1...         1023   
11  LEICA M11  /home/steven/git/camera-char/detector-noise/M1...         1023   
12  LEICA M11  /home/steven/git/camera-char/detector-noise/M1...         1023   

    white_level  width  height  iso      time       std         mean   min  \
0         16383   7424    4936   64  1.000000  0.967328  1022.963465  1001   
1         16383   7424    4936   64  0.500000  0.961643  1022.948249  1000   
2         16383   7424    4936   64  0.250000  0.958543  1023.282108  1001   
3         16383   7424    4936   64  0.125000  0.957108  1022.934205  1003   
4         16383   7424    4936   64  0.066667  0.957062  1022.930987  1003   
5         16383   7424    4936   64  0.033333  0.956962  1022.928360  1000   
6         16383   7424    4936   64  0.016667  0.956662  1022.930284   999   
7         16383   7424    4936   64  0.008000  0.956852  1023.273896  1001   
8         16383   7424    4936   64  0.004000  0.956960  1022.929697  1001   
9         16383   7424    4936   64  0.002000  0.956889  1022.923102  1000   
10        16383   7424    4936   64  0.001000  0.957387  1022.929390   998   
11        16383   7424    4936   64  0.000500  0.956601  1023.271485   999   
12        16383   7424    4936   64  0.000250  0.957294  1022.924316  1000   

     max         EV  
0   1050  13.954814  
1   1045  13.963317  
2   1048  13.967975  
3   1047  13.970137  
4   1045  13.970206  
5   1047  13.970357  
6   1045  13.970809  
7   1049  13.970523  
8   1046  13.970361  
9   1044  13.970467  
10  1045  13.969717  
11  1047  13.970902  
12  1049  13.969857

In [25]:
m11_60['std_2'] = m11_60['std']**2

In [26]:
px.line(m11_36[m11_36['iso'] == 3200], x='time', y='std', markers=True)

In [27]:
px.line(m11_36[m11_36['time'] == 0.001], x='iso', y='std', markers=True)

In [28]:
px.line(q3_36[q3_36['time'] == 0.001], x='iso', y='std', markers=True)

# Definition of ISO

ISO for a sensor is defined in two ways: either by an SNR target, or by the light power required to reach saturation of the detector with the applied gain. 

https://www.dxomark.com/glossary/iso-speed/
https://en.wikipedia.org/wiki/Lumen_(unit)

$H_{sat}$ is the exposure (in $lux \cdot s$) required to achieve saturation of the detector. Actually, it's usually the exposure to achieve 80% of the well capacity, to allow for some highlight recovery headroom. 

The ISO is defined by $S_{sat} = \frac{78}{H_{sat}}$



For example if $S_{sat}=100$, then $H_{sat} = 0.78 lux \cdot s$. 

$1 lux =1 lm/m^2$

Irradience is in $W/m^2$. Conversion from luminance to irradiance depends on the wavelength. 

For a given pixel size $p$ the total integrated power to reach saturation is 

This reference https://www.hhyu.org/posts/photons/ says 1 lumen is 4.09*10¹⁵ photons/s at the peak flux of 555 nm.

How many photos saturate the detector, and what is the final SNR?

$$
\begin{aligned}
H_{sat} &= \frac{78}{S_{sat}} \mathrm{lux} \cdot \mathrm{s} \\
&= \frac{78}{S_{sat}} \mathrm{lumen}/\mathrm{m}^2/\mathrm{s} \\
&= \frac{78}{S_{sat}} 4.09\times10^{15} \mathrm{photon}/\mathrm{m}^2 \\
&= \frac{3.19 \times10^{17}}{S_{sat}} \mathrm{photon}/\mathrm{m}^2 \\
\end{aligned}
$$

The Leica M11 has CMOS sensor, pixel pitch: 3.76 μm, 35 mm: 9528 x 6328 pixels (60.3 MP), so the total photons to saturate the
detector at a given ISO $S_{sat}$ is

In [29]:
iso = 50000
H_sat = 3.76e-6**2 * 3.19e17 / iso
(H_sat, 1/H_sat**0.5, np.log(H_sat**0.5)/np.log(2))

(90.197888, 0.10529356149799146, np.float64(3.2475108738979857))

The EV for a saturated detector is thus 7.7, but this seems way too low since the EV from the noise of the detector alone is as high as 13. Something is wrong. 

In [30]:
def iso_to_photon_saturation(iso, px, py, ):
    qe = 0.8 #guesstimate
    lumen = 4.09e15 #lumen in photons/s
    

In [31]:
import nbconvert

ModuleNotFoundError: No module named 'nbconvert'

In [ ]:
nbconvert.__version__

'6.5.4'